dm/worker/upgrade.go

```diff
@@ -0,0 +1,232 @@
+	defaultPreviousWorkerVersion = version{InternalNo: 0, ReleaseVersion: "None"}
+	// all versions exists in the history.
+	workerVersion1 = version{InternalNo: 1, ReleaseVersion: "v1.0.0-alpha"}
+)
+
+// The version of DM-worker used when upgrading from an older version.
+type version struct {
+	InternalNo     uint64 `json:"internal-no"`     // internal version number
+	ReleaseVersion string `json:"release-version"` // release version, like `v1.0.0`
+}
+
+// newVersion creates a new instance of version.
+func newVersion(internalNo uint64, releaseVersion string) version {
+	return version{
+		InternalNo:     internalNo,
+		ReleaseVersion: releaseVersion,
+	}
+}
+
+// newCurrentVersion creates a new instance of version match the current DM-worker.
+func newCurrentVersion() version {
+	return newVersion(currentWorkerInternalNo, utils.ReleaseVersion)
+}
+
+// compare compares the version with another version.
+// NOTE: also compare `ReleaseVersion` when needed.
+func (v *version) compare(other version) int {
+	if v.InternalNo < other.InternalNo {
+		return -1
+	} else if v.InternalNo == other.InternalNo {
+		return 0
+	}
+	return 1
+}
+
+// String implements Stringer.String.
+func (v version) String() string {
+	data, err := v.MarshalBinary()
+	if err != nil {
+		log.Errorf("[worker upgrade] marshal version (internal-no: %d, release-version: %s) to binary error %v",
+			v.InternalNo, v.ReleaseVersion, err)
+		return ""
+	}
+	return string(data)
+}
+
+// MarshalBinary implements encoding.BinaryMarshal.
+func (v *version) MarshalBinary() ([]byte, error) {
+	return json.Marshal(v)
+}
+
+// UnmarshalBinary implements encoding.BinaryMarshal.
+func (v *version) UnmarshalBinary(data []byte) error {
+	return json.Unmarshal(data, v)
+}
+
+// loadVersion loads the version of DM-worker from the levelDB.
+func loadVersion(h Getter) (ver version, err error) {
+	if whetherNil(h) {
+		return ver, errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := h.Get(dmWorkerVersionKey, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return defaultPreviousWorkerVersion, nil
+		}
+		return ver, errors.Annotatef(err, "load version with key %s from levelDB", dmWorkerVersionKey)
+	}
+	err = ver.UnmarshalBinary(data)
+	return ver, errors.Annotatef(err, "unmarshal version from data % X", data)
+}
+
+// saveVersion saves the version of DM-worker into the levelDB.
+func saveVersion(h Putter, ver version) error {
+	if whetherNil(h) {
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := ver.MarshalBinary()
+	if err != nil {
+		return errors.Annotatef(err, "marshal version %v to binary data", ver)
+	}
+
+	err = h.Put(dmWorkerVersionKey, data, nil)
+	return errors.Annotatef(err, "save version %v into levelDB with key %s", ver, dmWorkerVersionKey)
+}
+
+// tryUpgrade tries to upgrade from an older version.
+func tryUpgrade(dbDir string) error {
+	// 1. check the DB directory
+	fs, err := os.Stat(dbDir)
```

Comment:
```
notExist := false
fs, err := os.Stat(dbDir)
if err != nil {
	if os.IsNotExist(err) {
        notExist = true
	}
	return errors.Annotatef(err, "get stat for %s", dbDir)
} else if !fs.IsDir() { // should be a directory
	return errors.NotValidf("directory %s for DB", dbDir)
}

// 2. open the kv DB
db, err := openDB(dbDir, defaultKVConfig)
if err != nil {
	return errors.Annotatef(err, "open DB for %s", dbDir)
}
defer func() {
	err = db.Close()
	if err != nil {
		log.Errorf("[worker upgrade] close DB fail %v", err)
	}
}()

if notExist {
  ...
}
...
```

Reply:
done in https://github.com/pingcap/dm/pull/187/commits/4de40dee717712be8411e05d89c3cde50f441ee1.